### PR TITLE
feat: delay ack middleware

### PR DESCRIPTION
**Pull request by omritoptix**
# PR Standards

## Opening a pull request should be able to meet the following requirements:

---

For Author:

- [ ]  Targeted PR against correct branch
- [ ]  included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Linked to Github issue with discussion and accepted design
- [ ]  Targets only one github issue
- [ ]  Wrote unit and integration tests
- [ ]  All CI checks have passed
- [ ]  Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code)

---

For Reviewer:

- [ ]  confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Reviewers assigned
- [ ]  confirmed all author checklist items have been addressed

---;

After reviewer approval:

- [ ]  In case targets main branch, PR should be squashed and merged.
- [ ]  In case PR targets a release branch, PR should be rebased.

### CODE REVIEW

File: x/delayedack/ibc_middleware.go [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+) error {
+	return im.keeper.WriteAcknowledgement(ctx, chanCap, packet, ack)
+}
+
+// GetAppVersion returns the application version of the underlying application
+func (im IBCMiddleware) GetAppVersion(ctx sdk.Context, portID, channelID string) (string, bool) {
+	return im.keeper.GetAppVersion(ctx, portID, channelID)
+}
+
+/* ------------------------------- Custom Logic ------------------------------ */
```
Comment (mtsitrin) [inline x/delayedack/ibc_middleware.go:202]: move to separate`hooks.go` file

File: osmoutils/sumtree/README.md
Comment (mtsitrin) [inline osmoutils/sumtree/README.md:1]: not supposed to be here

File: proto/dymension/delayedack/genesis.proto [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+
+option go_package = "github.com/dymensionxyz/dymension/x/delayedack/types";
+
+// GenesisState defines the delayedack module's genesis state.
+message GenesisState {
+  Params params = 1 [(gogoproto.nullable) = false];
+}
+
+message RollappPacket {
```
Comment (mtsitrin) [inline proto/dymension/delayedack/genesis.proto:18]: not related to `genesis.proto`, probably move to `packet.proto`

File: proto/dymension/delayedack/query.proto
Comment (mtsitrin) [inline proto/dymension/delayedack/query.proto:1]: can be totally removed no? no params/queries

File: x/delayedack/ibc_middleware.go [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+		// Call the onRecvPacket callback for each packet
+		ack := im.app.OnRecvPacket(ctx, *rollappPacket.Packet, rollappPacket.Relayer)
+		// Write the acknowledgement to the chain only if it is synchronous
+		if ack != nil {
+			_, chanCap, err := im.keeper.LookupModuleByChannel(ctx, rollappPacket.Packet.DestinationPort, rollappPacket.Packet.DestinationChannel)
+			if err != nil {
+				logger.Error("Error looking up module by channel", "rollappID", rollappID, "error", err.Error())
+				return err
+			}
+			err = im.keeper.WriteAcknowledgement(ctx, chanCap, rollappPacket.Packet, ack)
```
Comment (mtsitrin) [inline x/delayedack/ibc_middleware.go:234]: why it's needed? isn't u just need to run `im.app.OnRecvPacket` as we skipped it previously in the middleware?

File: x/delayedack/keeper/rollapp_packet.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+	// Iterate over the range from lastProofHeight to proofHeight
+	iterator := sdk.KVStorePrefixIterator(store, prefix)
+	defer iterator.Close() // nolint: errcheck
+
+	for ; iterator.Valid(); iterator.Next() {
+		var val types.RollappPacket
+		k.cdc.MustUnmarshal(iterator.Value(), &val)
+		keyParts := strings.Split(string(iterator.Key()), "/")
+		packetProofHeight, _ := strconv.ParseUint(keyParts[2], 10, 64)
```
Comment (mtsitrin) [inline x/delayedack/keeper/rollapp_packet.go:77]: check len(keyParts) before

File: x/delayedack/types/errors.go [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+
+// DONTCOVER
+
+import (
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+)
+
+// x/delayedack module sentinel errors
+var (
+	ErrSample = sdkerrors.Register(ModuleName, 1100, "sample error")
```
Comment (mtsitrin) [inline x/delayedack/types/errors.go:11]: remove, or create dedicated errors

File: x/delayedack/types/key_rollapp_packet.go [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+var _ binary.ByteOrder
+
+const (
+	// RollappPacketKeyPrefix is the prefix to retrieve all RollappPackets
+	RollappPacketKeyPrefix = "RollappPacket/value/"
+)
+
+// GetRollappPacketKey constructs a key for a specific RollappPacket
+func GetRollappPacketKey(
```
Comment (mtsitrin) [inline x/delayedack/types/key_rollapp_packet.go:18]: why u use height in the key path? it's not unique
why not using just the packet's `Sequence`?

File: x/delayedack/ibc_middleware.go (hunk 3 above)
Comment (omritoptix) [reply x/delayedack/ibc_middleware.go:234]: If the `OnRecvPacket` returns an ack (i.e sync acknowledgement) then you don’t need it, but if it returns nil (async ack), then you need to call it by yourself once the ack is received.

File: x/delayedack/types/key_rollapp_packet.go (hunk 6 above)
Comment (omritoptix) [reply x/delayedack/types/key_rollapp_packet.go:18]: I need to iterate and get the relevant packets by heights in an efficient way once it's related state is finalized (look at `ListRollappPackets`). 
the overall rollapp packet key is unique as it also combines sequence and channel.

File: x/delayedack/ibc_middleware.go [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+		// Update the packet status
+		im.keeper.UpdateRollappPacketStatus(ctx, rollappID, rollappPacket, types.RollappPacket_ACCEPTED)
+		// Call the onRecvPacket callback for each packet
+		ack := im.app.OnRecvPacket(ctx, *rollappPacket.Packet, rollappPacket.Relayer)
+		// Write the acknowledgement to the chain only if it is synchronous
+		if ack != nil {
+			_, chanCap, err := im.keeper.LookupModuleByChannel(ctx, rollappPacket.Packet.DestinationPort, rollappPacket.Packet.DestinationChannel)
+			if err != nil {
+				logger.Error("Error looking up module by channel", "rollappID", rollappID, "error", err.Error())
+				return err
```
Comment (mtsitrin) [inline x/delayedack/ibc_middleware.go:232]: u probably want to continue with other packets incase of error on specific packet

File: x/delayedack/ibc_middleware.go [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+			}
+			err = im.keeper.WriteAcknowledgement(ctx, chanCap, rollappPacket.Packet, ack)
+			if err != nil {
```
Comment (mtsitrin) [inline x/delayedack/ibc_middleware.go:235]: if the packets will fail here, it won't retry as it's already `RollappPacket_ACCEPTED`?
desired outcome?

File: x/delayedack/keeper/rollapp_packet.go [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+	prefix = append(prefix, []byte("/")...)
+
+	// Iterate over the range from lastProofHeight to proofHeight
+	iterator := sdk.KVStorePrefixIterator(store, prefix)
+	defer iterator.Close() // nolint: errcheck
+
+	for ; iterator.Valid(); iterator.Next() {
+		var val types.RollappPacket
+		k.cdc.MustUnmarshal(iterator.Value(), &val)
+		keyParts := strings.Split(string(iterator.Key()), "/")
```
Comment (mtsitrin) [inline x/delayedack/keeper/rollapp_packet.go:76]: why u need to extract height it from the key? u have it in the value

File: x/delayedack/ibc_middleware.go (hunk 7 above)
Comment (omritoptix) [reply x/delayedack/ibc_middleware.go:232]: good catch!

File: x/delayedack/ibc_middleware.go (hunk 8 above)
Comment (omritoptix) [reply x/delayedack/ibc_middleware.go:235]: yea similiar to how vanilla works. if writeAck fails it returns an error and doesn't retry.
This packet will then get timeout at somepoint. 
this packet can be resent again for retrying.
However I shouldn't return an err here, just log it here (similar to your previous comments)

File: x/delayedack/keeper/rollapp_packet.go (hunk 9 above)
Comment (omritoptix) [reply x/delayedack/keeper/rollapp_packet.go:76]: true!

**Comment by mtsitrin** [comment]
pls add guthub issue for UT this middleware
probably dependent on the infra work that needs to be done in https://github.com/dymensionxyz/dymension/issues/312

**Comment by omritoptix** [comment]
yea already added it few days ago https://github.com/dymensionxyz/dymension/issues/372

File: proto/dymension/delayedack/query.proto
Comment (omritoptix) [reply proto/dymension/delayedack/query.proto:1]: it starts breaking the ModuleBasics interface. can be done but requires more work which I don't want to put currently. I'll add it as technical debt.

File: x/delayedack/ibc_middleware.go [hunk 10]
```diff
@@ -228,17 +218,15 @@ func (im IBCMiddleware) FinalizeRollappPackets(ctx sdk.Context, rollappID string
 		if ack != nil {
 			_, chanCap, err := im.keeper.LookupModuleByChannel(ctx, rollappPacket.Packet.DestinationPort, rollappPacket.Packet.DestinationChannel)
 			if err != nil {
-				logger.Error("Error looking up module by channel", "rollappID", rollappID, "error", err.Error())
-				return err
+				logger.Error("Error looking up module by channel", "rollappID", rollappID, "error", "sequence", rollappPacket.Packet.GetSequence(), "destination channel", rollappPacket.Packet.GetDestChannel(), err.Error())
```
Comment (mtsitrin) [inline x/delayedack/ibc_middleware.go:221]: misplaced `err.Error()`

File: x/delayedack/ibc_middleware.go [hunk 11]
```diff
@@ -201,25 +201,15 @@ func (im IBCMiddleware) GetAppVersion(ctx sdk.Context, portID, channelID string)
...
-		return err
-	}
-
-	return nil
-}
-
 // FinalizeRollappPackets finalizes the packets for the given rollapp until the given height which is
 // the end height of the latest finalized state
-func (im IBCMiddleware) FinalizeRollappPackets(ctx sdk.Context, rollappID string, stateEndHeight uint64) error {
+func (im IBCMiddleware) FinalizeRollappPackets(ctx sdk.Context, rollappID string, stateEndHeight uint64) {
```
Comment (mtsitrin) [inline x/delayedack/ibc_middleware.go:206]: probably move to hooks as well

**Review by mtsitrin — APPROVED** [review]

**Pull request merged**